### PR TITLE
Add a `min_type_alias_impl_trait` feature gate

### PR DESCRIPTION
**Pull request by oli-obk**
This new feature gate only permits type alias impl trait to be constrained by function and trait method return types. All other possible constraining sites like const/static types, closure return types and binding types are now forbidden and gated under the `type_alias_impl_trait` and `impl_trait_in_bindings` feature gates (which are both marked as incomplete, as they have various ways to ICE the compiler or cause query cycles where they shouldn't).

r? @nikomatsakis 

This is best reviewed commit-by-commit

### CODE REVIEW

File: compiler/rustc_typeck/src/check/check.rs [hunk 1]
```diff
@@ -87,8 +87,26 @@ pub(super) fn check_fn<'a, 'tcx>(
 
     let declared_ret_ty = fn_sig.output();
 
-    let revealed_ret_ty =
-        fcx.instantiate_opaque_types_from_value(fn_id, declared_ret_ty, decl.output.span());
+    let feature = match tcx.hir().get(fn_id) {
+        Node::Item(hir::Item { kind: ItemKind::Fn(..), .. }) |
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/check.rs:91]: I'd like to see comments on each arm with an example of the corresponding rust code

File: compiler/rustc_typeck/src/check/check.rs [hunk 2]
```diff
@@ -87,8 +87,26 @@ pub(super) fn check_fn<'a, 'tcx>(
...
+        Node::ImplItem(hir::ImplItem {
+            kind: hir::ImplItemKind::Fn(..), ..
+        }) => None,
+        // I don't know if TAIT uses in trait declarations make sense at all
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/check.rs:95]: I'd probably just drop this comment -- save it for the RFC thread, @oli-obk! =)

PS this is my silly way of telling you I think they do ;)

**Review by nikomatsakis — APPROVED** [review]
This PR...seems pretty clean, @oli-obk! Left one suggestion. r=me with the comments.

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/83146#issuecomment-799486325)

@bors r-

**Comment by oli-obk** [comment]
@bors rollup=never

could theoretically affect perf

**Pull request merged**